Src/FluentAssertions/Formatting/FormattedObjectGraph.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using FluentAssertions.Execution;
+
+namespace FluentAssertions.Formatting
+{
+    /// <summary>
+    /// This class is used by the <see cref="Formatter"/> class to collect all the output of the (nested calls of a) <see cref="IValueFormatter"/> into
+    /// a the final representation.
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="FormattedObjectGraph"/> will ensure that the number of lines will be limited
+    /// to the maximum lines provided through its constructor. It will throw
+    /// a <see cref="MaxLinesExceededException"/> when the maximum has been reached.
+    /// </remarks>
+    public class FormattedObjectGraph
+    {
+        private readonly int maxLines;
+        private readonly List<string> lines = new();
+        private readonly StringBuilder lineBuilder = new();
+        private int indentation;
+        private string lineBuilderWhitespace = string.Empty;
+
+        public FormattedObjectGraph(int maxLines)
+        {
+            this.maxLines = maxLines;
+        }
+
+        /// <summary>
+        /// The number of spaces that should be used by every indentation level.
+        /// </summary>
+        public static int SpacesPerIndentation => 4;
+
+        /// <summary>
+        /// Returns the number of lines of text currently in the graph.
+        /// </summary>
+        public int LineCount => lines.Count + ((lineBuilder.Length > 0) ? 1 : 0);
+
+        /// <summary>
+        /// Starts a new line with the provided text fragment. Additional text can be added to
+        /// that same line through <see cref="AddFragment"/>.
+        /// </summary>
+        public void AddFragmentOnNewLine(string fragment)
+        {
+            FlushCurrentLine();
+
+            AddFragment(fragment);
+        }
+
+        /// <summary>
+        /// Starts a new line with the provided line of text that does not allow
+        /// adding more fragments of text.
+        /// </summary>
+        public void AddLine(string line)
+        {
+            FlushCurrentLine();
+
+            AppendSafely(Whitespace + line);
+        }
+
+        /// <summary>
+        /// Starts or continues the current line with additional fragments of text.
```

Comment:
❓ I'm not sure what the additional fragments in the plural "fragments" refer to.

Reply:
I rephrased it. Hopefully it clarifies it a bit better now.